sticker/src/encoder.rs

```diff
@@ -135,3 +136,124 @@ impl SentenceEncoder for LayerEncoder {
         Ok(encoding)
     }
 }
+
+/// An encoder wrapper that encodes/decodes to a categorical label.
+pub struct CategoricalEncoder<E, V>
+where
+    V: Eq + Hash,
+{
+    inner: E,
+    numberer: Numberer<V>,
+}
+
+impl<E, V> CategoricalEncoder<E, V>
+where
+    V: Eq + Hash,
+{
+    pub fn new(encoder: E, numberer: Numberer<V>) -> Self {
+        CategoricalEncoder {
+            inner: encoder,
+            numberer,
+        }
+    }
+}
+
+impl<E> SentenceEncoder for CategoricalEncoder<E, E::Encoding>
+where
+    E: SentenceEncoder,
+    E::Encoding: Clone + Eq + Hash,
+{
+    type Encoding = usize;
+
+    fn encode(&mut self, sentence: &Sentence) -> Result<Vec<Self::Encoding>, Error> {
+        let encoding = self.inner.encode(sentence)?;
+        let categorical_encoding = encoding.into_iter().map(|e| self.numberer.add(e)).collect();
```

Comment:
Fixing #28 would require adding a flag or another method here, e.g.
```Rust
fn encode(&mut self, sentence: &Sentence, mode: Mode) -> Result<Vec<Self::Encoding>, Error> {
    let encoding = self.inner.encode(sentence)?;
    let categorical_encoding = match mode {
        Mode::Train => encoding.into_iter().map(|e| self.numberer.add(e)).collect()
        // would require starting the numberer at 2
        Mode::Validate => encoding.into_iter().map(|e| self.numberer.number(e).unwrap_or(1)).collect()
}
```

Reply:
I don't like adding arguments. I think this should rather be addressed by the `Numberer`, which could be thawed/frozen. (As dpar does it.)